File: skills/agent-swarm/scripts/backends/acp/adapter.py
"""Execution Backend adapter controlling detached ACP Workers by Launch ID."""

import json
import os
import pathlib
import secrets
import subprocess
import sys
import time

import execution_state
import state_store
from backends.base import (
    AgentBackend,
    BackendPendingError,
    BackendUnknownError,
    ObserveResult,
    SpawnRequest,
    SpawnResult,
    StopRequest,
)
from backends.acp.processes import pid_alive, process_group_alive, terminate_process_group
from backends.acp.worker_protocol import control_request


WORKER = pathlib.Path(__file__).resolve().with_name("worker.py")
DETACHED_LAUNCHER = pathlib.Path(__file__).resolve().with_name("detached_launcher.py")
MIN_FRESH_WORKER_LAUNCH_TIMEOUT_SECONDS = 1.0


class AcpBackend(AgentBackend):
    backend_id = "acp"

    def __init__(self, config=None, execution_record=None):
        self.config = dict(config or {})
        self.execution_record = dict(execution_record or {})

    def _record(self, launch_id=None):
        launch_id = launch_id or self.execution_record.get("launch_id")
        record = state_store.get_launch(launch_id) if launch_id else None
        if record is None:
            raise RuntimeError("ACP Launch record not found")
        if record["backend_id"] != "acp":
            raise RuntimeError("Launch backend is not ACP")
        self.execution_record = record
        return record

    @staticmethod
    def _job_id(record):
        return record.get("backend_ref") or "acp-launch:%s" % record["launch_id"]

    def _ping(self, record, timeout=0.5):
        endpoint = record.get("control_endpoint")
        if not endpoint or not pathlib.Path(endpoint).exists():
            return None
        try:
            result = control_request(
                endpoint,
                "ping",
                {"launch_id": record["launch_id"]},
                timeout=timeout,
            )
        except Exception:
            return None
        if result.get("ok") and int(result.get("launch_id", -1)) == record["launch_id"]:
            return result
        return None

    def _launch_worker(self, request, record):
        runtime_root = state_store.runtime_root()
        log_path = (
            runtime_root
            / "logs"
            / record["root_id"]
            / "acp"
            / ("launch-%s.worker.log" % record["launch_id"])
        )
        log_path.parent.mkdir(parents=True, exist_ok=True, mode=0o700)
        descriptor = os.open(str(log_path), os.O_WRONLY | os.O_CREAT | os.O_APPEND, 0o600)
        candidate_nonce = secrets.token_hex(16)
        environment = {
            key: value for key, value in os.environ.items() if not key.startswith("AGENT_SWARM_")
        }
        environment["AGENT_SWARM_HOME"] = str(runtime_root)
        environment["AGENT_SWARM_EXECUTION_NONCE"] = candidate_nonce
        environment["PYTHONDONTWRITEBYTECODE"] = "1"
        try:
            subprocess.run(
                [
                    sys.executable,
                    str(DETACHED_LAUNCHER),
                    sys.executable,
                    str(WORKER),
                    "--launch-id",
                    str(record["launch_id"]),
                    "--candidate-nonce",
                    candidate_nonce,
                ],
                cwd=request.cwd,
                env=environment,
                stdin=subprocess.DEVNULL,
                stdout=descriptor,
                stderr=descriptor,
                close_fds=True,
                check=True,
            )
        finally:
            os.close(descriptor)

    def _advance_absent_launch(self, record):
        """Close one proven-absent Launch and append its retry Launch."""
        if record["status"] != "starting" or record.get("stop_requested_at") is not None:
            return False
        if process_group_alive(record.get("worker_pid")) or process_group_alive(record.get("agent_pid")):
            return False
        endpoint = record.get("control_endpoint")
        if endpoint:
            try:
                pathlib.Path(endpoint).unlink()
            except FileNotFoundError:
                pass
            except OSError as exc:
                raise BackendUnknownError(
                    "ACP control endpoint could not be removed before Launch retry"
                ) from exc
        with state_store.transaction() as con:
            current = state_store.get_launch(record["launch_id"], con)
            latest = state_store.get_current_launch(record["attempt_id"], con)
            attempt = state_store.get_attempt(record["attempt_id"], con)
            if not current or not latest or not attempt:
                return False
            if (
                latest["launch_id"] != current["launch_id"]
                or current.get("owner_nonce") != record.get("owner_nonce")
                or current["status"] != "starting"
                or current.get("stop_requested_at") is not None
                or attempt["state"] != "assigned"
            ):
                return False
            timestamp = state_store.now()
            con.execute(
                """UPDATE launches SET status='closed', prompt_state='cancelled',
                     exit_reason='worker_agent_control_absent', closed_at=?, last_event_at=?
                   WHERE launch_id=? AND status='starting'""",
                (timestamp, timestamp, current["launch_id"]),
            )
            cursor = con.execute(
                """INSERT INTO launches(
                     attempt_id, launch_no, session_name, status, prompt_state,
                     created_at, last_event_at
                   ) VALUES (?, ?, ?, 'starting', 'pending', ?, ?)""",
                (
                    current["attempt_id"],
                    current["launch_no"] + 1,
                    current["session_name"],
                    timestamp,
                    timestamp,
                ),
            )
            launch_id = cursor.lastrowid
            payload = {
                "root_id": current["root_id"],
                "task_id": current["task_id"],
                "attempt_id": current["attempt_id"],
                "launch_id": launch_id,
                "backend_id": "acp",
            }
            con.execute(
                """INSERT INTO effects(
                     root_id, attempt_id, launch_id, effect_type, payload_json,
                     idempotency_key, status, attempts, created_at
                   ) VALUES (?, ?, ?, 'spawn_agent', ?, ?, 'pending', 0, ?)""",
                (
                    current["root_id"],
                    current["attempt_id"],
                    launch_id,
                    json.dumps(payload, ensure_ascii=False, sort_keys=True),
                    "spawn:%s" % launch_id,
                    timestamp,
                ),
            )
            state_store.append_event(
                con,
                current["root_id"],
                "LaunchRetried",
                {
                    "previous_launch_id": current["launch_id"],
                    "launch_id": launch_id,
                    "reason": "worker_agent_control_absent",
                },
                task_id=current["task_id"],
                attempt_id=current["attempt_id"],
            )
        self.execution_record = state_store.get_launch(launch_id)
        return True

    def spawn(self, request):
        if not isinstance(request, SpawnRequest):
            raise TypeError("ACP spawn requires SpawnRequest")
        try:
            launch_id = int(request.metadata.get("launch_id"))
        except (TypeError, ValueError) as exc:
            raise RuntimeError("ACP spawn requires an integer launch_id") from exc
        record = self._record(launch_id)
        ping = self._ping(record)
        if record.get("ready_at") is not None and (ping is not None or record["status"] == "closed"):
            session = state_store.get_session_for_launch(record["launch_id"])
            return SpawnResult(
                job_id=self._job_id(record),
                session_name=record["session_name"],
                extras={
                    "launch_id": record["launch_id"],
                    "worker_pid": record.get("worker_pid"),
                    "agent_pid": record.get("agent_pid"),
                    "external_session_id": session.get("external_session_id") if session else None,
                    "protocol_version": session.get("protocol_version") if session else None,
                },
            )
        if record["status"] == "closed":
            raise RuntimeError(record.get("exit_reason") or "ACP Worker closed before ready")
        worker_alive = pid_alive(record.get("worker_pid"))
        agent_alive = pid_alive(record.get("agent_pid"))
        if record.get("owner_nonce") and not worker_alive and agent_alive:
            raise BackendUnknownError("ACP Worker is absent while its Agent Process is still alive")
        launched_worker = record.get("owner_nonce") is None
        if launched_worker:
            self._launch_worker(request, record)

        timeout = float(self.config.get("worker_launch_timeout_seconds") or 12)
        if launched_worker:
            timeout = max(timeout, MIN_FRESH_WORKER_LAUNCH_TIMEOUT_SECONDS)
        deadline = time.monotonic() + timeout
        while time.monotonic() < deadline:
            record = self._record(launch_id)
            ping = self._ping(record)
            if record.get("ready_at") is not None and (ping is not None or record["status"] == "closed"):
                session = state_store.get_session_for_launch(record["launch_id"])
                return SpawnResult(
                    job_id=self._job_id(record),
                    session_name=record["session_name"],
                    extras={
                        "launch_id": record["launch_id"],
                        "worker_pid": record.get("worker_pid"),
                        "agent_pid": record.get("agent_pid"),
                        "external_session_id": session.get("external_session_id") if session else None,
                        "protocol_version": session.get("protocol_version") if session else None,
                    },
                )
            if record["status"] in {"error", "turn_ended", "closed"}:
                raise RuntimeError(record.get("exit_reason") or "ACP Worker failed before ready")
            if record.get("owner_nonce") and not pid_alive(record.get("worker_pid")):
                if pid_alive(record.get("agent_pid")):
                    raise BackendUnknownError(
                        "ACP Worker exited before ready while Agent Process remains alive"
                    )
            time.sleep(0.03)
        record = self._record(launch_id)
        if self._advance_absent_launch(record):
            raise BackendPendingError("absent ACP Launch was fenced; replacement Launch appended")
        raise BackendPendingError("ACP Worker is still starting")

    def stop(self, request):
        if not isinstance(request, StopRequest):
            raise TypeError("ACP stop requires StopRequest")
        record = self._record()
        execution_state.request_stop(record["launch_id"])
        record = self._record()
        endpoint = record.get("control_endpoint")
        if endpoint and pathlib.Path(endpoint).exists():
            try:
                control_request(
                    endpoint,
                    "stop",
                    {"launch_id": record["launch_id"], "timeout": 8},
                    timeout=10,
                )
            except Exception:
                pass
        deadline = time.monotonic() + 4
        while time.monotonic() < deadline:
            record = self._record()
            endpoint_exists = bool(
                record.get("control_endpoint") and pathlib.Path(record["control_endpoint"]).exists()
            )
            if (
                not process_group_alive(record.get("worker_pid"))
                and not process_group_alive(record.get("agent_pid"))
                and not endpoint_exists
            ):
                self._close_record(record, "stopped")
                return {"stopped": True}
            time.sleep(0.05)
        nonce = record.get("owner_nonce")
        agent_clean = terminate_process_group(record.get("agent_pid"), grace=0.5, expected_nonce=nonce)
        worker_clean = terminate_process_group(record.get("worker_pid"), grace=0.5, expected_nonce=nonce)
        endpoint = record.get("control_endpoint")
        if endpoint and not pid_alive(record.get("worker_pid")):
            try:
                pathlib.Path(endpoint).unlink()
            except FileNotFoundError:
                pass
        endpoint_clean = not (endpoint and pathlib.Path(endpoint).exists())
        if agent_clean and worker_clean and endpoint_clean:
            self._close_record(record, "forced_stop")
            return {"stopped": True, "forced": True}
        raise BackendUnknownError("ACP stop could not prove Worker/Agent cleanup")

    @staticmethod
    def _close_record(record, reason):
        with state_store.transaction() as con:
            timestamp = state_store.now()
            con.execute(
                """UPDATE launches SET status='closed', prompt_state='cancelled',
                     exit_reason=COALESCE(exit_reason, ?), closed_at=COALESCE(closed_at, ?),
                     last_event_at=? WHERE launch_id=?""",
                (reason, timestamp, timestamp, record["launch_id"]),
            )
            con.execute(
                """UPDATE acp_sessions SET status='closed', closed_at=COALESCE(closed_at, ?)
                   WHERE launch_id=? AND status='active'""",
                (timestamp, record["launch_id"]),
            )

    def observe(self, *, job_id=None, session_name=None, cwd=None):
        record = self._record()
        if job_id and job_id not in {record.get("backend_ref"), self._job_id(record)}:
            return ObserveResult("unknown", error="job id does not match Launch")
        worker_alive = pid_alive(record.get("worker_pid"))
        agent_alive = pid_alive(record.get("agent_pid"))
        worker_group_alive = process_group_alive(record.get("worker_pid"))
        agent_group_alive = process_group_alive(record.get("agent_pid"))
        endpoint_exists = bool(
            record.get("control_endpoint") and pathlib.Path(record["control_endpoint"]).exists()
        )
        ping = self._ping(record) if endpoint_exists else None
        if ping is not None and worker_alive:
            return ObserveResult("present", session=ping)
        if not worker_alive and (agent_alive or agent_group_alive):
            return ObserveResult(
                "unknown",
                session={"agent_pid": record.get("agent_pid")},
                error="orphan Agent Process is alive without its ACP Worker",
            )
        if not worker_group_alive and not agent_group_alive and not endpoint_exists:
            return ObserveResult("absent")
        return ObserveResult("unknown", error="ACP Launch facts are contradictory")

    def list_sessions(self, *, cwd=None):
        if not self.execution_record.get("root_id"):
            return []
        return [
            {
                "id": self._job_id(row),
                "job_id": self._job_id(row),
                "launch_id": row["launch_id"],
                "name": row["session_name"],
                "session_name": row["session_name"],
                "state": row["status"],
                "worker_pid": row.get("worker_pid"),
                "agent_pid": row.get("agent_pid"),
            }
            for row in state_store.list_launches(self.execution_record["root_id"])
            if row["backend_id"] == "acp" and row["status"] not in {"closed", "error", "turn_ended"}
        ]

    def supports_hooks(self):
        return False

File: skills/agents-orchestrator/scripts/backends/acp/adapter.ts
import { closeSync, constants as fsConstants, existsSync, mkdirSync, openSync, unlinkSync } from "node:fs";
import { randomBytes } from "node:crypto";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import * as executionState from "../../execution_state.ts";
import * as stateStore from "../../state_store.ts";
import { canonicalJson, RuntimeError, type RuntimeRecord } from "../../runtime_types.ts";
import {
  AgentBackend,
  BackendPendingError,
  BackendUnknownError,
  ObserveResult,
  SpawnRequest,
  SpawnResult,
  StopRequest,
} from "../base.ts";
import { pidAlive, processGroupAlive, terminateProcessGroup } from "./processes.ts";
import { controlRequestSync } from "./worker_protocol.ts";

const WORKER = fileURLToPath(new URL("./worker.ts", import.meta.url));
const MIN_FRESH_WORKER_LAUNCH_TIMEOUT_SECONDS = 1;

function integer(value: unknown): number | null {
  const parsed = Number(value);
  return Number.isSafeInteger(parsed) && parsed > 0 ? parsed : null;
}

function unlinkMissingOkay(path: string): void {
  try { unlinkSync(path); } catch (error) { if ((error as NodeJS.ErrnoException).code !== "ENOENT") throw error; }
}

export class AcpBackend extends AgentBackend {
  readonly backendId = "acp";
  private config: RuntimeRecord;
  private executionRecord: RuntimeRecord;

  constructor(config: RuntimeRecord = {}, executionRecord: RuntimeRecord = {}) {
    super();
    this.config = { ...config };
    this.executionRecord = { ...executionRecord };
  }

  private record(launchId?: number | null): RuntimeRecord {
    const id = launchId ?? integer(this.executionRecord.launch_id);
    const record = id ? stateStore.getLaunch(id) : null;
    if (!record) throw new RuntimeError("ACP Launch record not found");
    if (record.backend_id !== "acp") throw new RuntimeError("Launch backend is not ACP");
    this.executionRecord = record;
    return record;
  }

  private static jobId(record: RuntimeRecord): string {
    return typeof record.backend_ref === "string" && record.backend_ref
      ? record.backend_ref : `acp-launch:${record.launch_id}`;
  }

  private ping(record: RuntimeRecord, timeoutSeconds = 0.5): RuntimeRecord | null {
    const endpoint = record.control_endpoint;
    if (typeof endpoint !== "string" || !existsSync(endpoint)) return null;
    try {
      const result = controlRequestSync(endpoint, "ping", { launch_id: record.launch_id }, timeoutSeconds);
      return result.ok && Number(result.launch_id) === Number(record.launch_id) ? result : null;
    } catch { return null; }
  }

  private launchWorker(request: SpawnRequest, record: RuntimeRecord): void {
    const logPath = join(
      stateStore.runtimeRoot(), "logs", String(record.root_id), "acp", `launch-${record.launch_id}.worker.log`,
    );
    mkdirSync(dirname(logPath), { recursive: true, mode: 0o700 });
    const descriptor = openSync(logPath, fsConstants.O_WRONLY | fsConstants.O_CREAT | fsConstants.O_APPEND, 0o600);
    const candidateNonce = randomBytes(16).toString("hex");
    const environment: Record<string, string> = {};
    const scrubbedSuffixes = new Set(["ROOT_ID", "TASK_ID", "ATTEMPT_ID", "ACTOR_TOKEN", "AGENT_ID", "EXECUTION_NONCE", "SKILL_DIR"]);
    for (const [key, value] of Object.entries(process.env)) {
      const suffix = key.startsWith("AGENTS_ORCHESTRATOR_") ? key.slice("AGENTS_ORCHESTRATOR_".length) : null;
      if (value !== undefined && !(suffix && scrubbedSuffixes.has(suffix))) environment[key] = value;
    }
    environment.AGENTS_ORCHESTRATOR_HOME = stateStore.runtimeRoot();
    environment.AGENTS_ORCHESTRATOR_EXECUTION_NONCE = candidateNonce;
    let child: Bun.Subprocess;
    try {
      child = Bun.spawn({
        cmd: [process.execPath, WORKER, "--launch-id", String(record.launch_id), "--candidate-nonce", candidateNonce],
        cwd: request.cwd,
        env: environment,
        stdin: "ignore",
        stdout: descriptor,
        stderr: descriptor,
        detached: true,
      });
      child.unref();
    } finally { closeSync(descriptor); }
  }

  private advanceAbsentLaunch(record: RuntimeRecord): boolean {
    if (record.status !== "starting" || record.stop_requested_at !== null) return false;
    if (processGroupAlive(record.worker_pid) || processGroupAlive(record.agent_pid)) return false;
    const endpoint = record.control_endpoint;
    if (typeof endpoint === "string") {
      try { unlinkMissingOkay(endpoint); }
      catch { throw new BackendUnknownError("ACP control endpoint could not be removed before Launch retry"); }
    }
    let nextLaunchId: number | null = null;
    stateStore.transaction((connection) => {
      const current = stateStore.getLaunch(Number(record.launch_id), connection);
      const latest = current ? stateStore.getCurrentLaunch(Number(current.attempt_id), connection) : null;
      const attempt = current ? stateStore.getAttempt(Number(current.attempt_id), connection) : null;
      if (!current || !latest || !attempt || Number(latest.launch_id) !== Number(current.launch_id) ||
          current.owner_nonce !== record.owner_nonce || current.status !== "starting" ||
          current.stop_requested_at !== null || attempt.state !== "assigned") return;
      const timestamp = stateStore.now();
      connection.execute(
        `UPDATE launches SET status='closed', prompt_state='cancelled',
            exit_reason='worker_agent_control_absent', closed_at=?, last_event_at=?
          WHERE launch_id=? AND status='starting'`,
        [timestamp, timestamp, current.launch_id],
      );
      const cursor = connection.execute(
        `INSERT INTO launches(attempt_id, launch_no, session_name, status, prompt_state, created_at, last_event_at)
         VALUES (?, ?, ?, 'starting', 'pending', ?, ?)`,
        [current.attempt_id, Number(current.launch_no) + 1, current.session_name, timestamp, timestamp],
      );
      nextLaunchId = Number(cursor.lastrowid);
      const payload = {
        root_id: current.root_id, task_id: current.task_id, attempt_id: current.attempt_id,
        launch_id: nextLaunchId, backend_id: "acp",
      };
      connection.execute(
        `INSERT INTO effects(root_id, attempt_id, launch_id, effect_type, payload_json,
          idempotency_key, status, attempts, created_at)
         VALUES (?, ?, ?, 'spawn_agent', ?, ?, 'pending', 0, ?)`,
        [current.root_id, current.attempt_id, nextLaunchId, canonicalJson(payload), `spawn:${nextLaunchId}`, timestamp],
      );
      stateStore.appendEvent(
        connection, String(current.root_id), "LaunchRetried",
        { previous_launch_id: current.launch_id, launch_id: nextLaunchId, reason: "worker_agent_control_absent" },
        Number(current.task_id), Number(current.attempt_id),
      );
    });
    if (nextLaunchId === null) return false;
    this.executionRecord = stateStore.getLaunch(nextLaunchId)!;
    return true;
  }

  spawn(request: SpawnRequest): SpawnResult {
    if (!(request instanceof SpawnRequest)) throw new TypeError("ACP spawn requires SpawnRequest");
    const launchId = integer(request.metadata.launch_id);
    if (!launchId) throw new RuntimeError("ACP spawn requires an integer launch_id");
    let record = this.record(launchId);
    let ping = this.ping(record);
    if (record.ready_at !== null && (ping || record.status === "closed")) return this.spawnResult(record);
    if (record.status === "closed") throw new RuntimeError(String(record.exit_reason || "ACP Worker closed before ready"));
    const workerAlive = pidAlive(record.worker_pid);
    const agentAlive = pidAlive(record.agent_pid);
    if (record.owner_nonce && !workerAlive && agentAlive) throw new BackendUnknownError("ACP Worker is absent while its Agent Process is still alive");
    const launchedWorker = !record.owner_nonce;
    if (launchedWorker) this.launchWorker(request, record);
    let timeout = Number(this.config.worker_launch_timeout_seconds || 12);
    if (launchedWorker) timeout = Math.max(timeout, MIN_FRESH_WORKER_LAUNCH_TIMEOUT_SECONDS);
    const deadline = performance.now() + timeout * 1000;
    while (performance.now() < deadline) {
      record = this.record(launchId);
      ping = this.ping(record);
      if (record.ready_at !== null && (ping || record.status === "closed")) return this.spawnResult(record);
      if (new Set(["error", "turn_ended", "closed"]).has(String(record.status))) {
        throw new RuntimeError(String(record.exit_reason || "ACP Worker failed before ready"));
      }
      if (record.owner_nonce && !pidAlive(record.worker_pid) && pidAlive(record.agent_pid)) {
        throw new BackendUnknownError("ACP Worker exited before ready while Agent Process remains alive");
      }
      Bun.sleepSync(30);
    }
    record = this.record(launchId);
    if (this.advanceAbsentLaunch(record)) throw new BackendPendingError("absent ACP Launch was fenced; replacement Launch appended");
    throw new BackendPendingError("ACP Worker is still starting");
  }

  private spawnResult(record: RuntimeRecord): SpawnResult {
    const session = stateStore.getSessionForLaunch(Number(record.launch_id));
    return new SpawnResult(AcpBackend.jobId(record), String(record.session_name), {
      launch_id: record.launch_id,
      worker_pid: record.worker_pid,
      agent_pid: record.agent_pid,
      external_session_id: session?.external_session_id ?? null,
      protocol_version: session?.protocol_version ?? null,
    });
  }

  stop(request: StopRequest): RuntimeRecord {
    if (!(request instanceof StopRequest)) throw new TypeError("ACP stop requires StopRequest");
    let record = this.record();
    executionState.requestStop(Number(record.launch_id));
    record = this.record();
    const endpoint = record.control_endpoint;
    if (typeof endpoint === "string" && existsSync(endpoint)) {
      try { controlRequestSync(endpoint, "stop", { launch_id: record.launch_id, timeout: 8 }, 10); } catch { /* bounded fallback below */ }
    }
    const deadline = performance.now() + 4000;
    while (performance.now() < deadline) {
      record = this.record();
      const endpointExists = typeof record.control_endpoint === "string" && existsSync(record.control_endpoint);
      if (!processGroupAlive(record.worker_pid) && !processGroupAlive(record.agent_pid) && !endpointExists) {
        this.closeRecord(record, "stopped");
        return { stopped: true };
      }
      Bun.sleepSync(50);
    }
    const nonce = typeof record.owner_nonce === "string" ? record.owner_nonce : null;
    const agentClean = terminateProcessGroup(record.agent_pid, { graceSeconds: 0.5, expectedNonce: nonce });
    const workerClean = terminateProcessGroup(record.worker_pid, { graceSeconds: 0.5, expectedNonce: nonce });
    if (typeof record.control_endpoint === "string" && !pidAlive(record.worker_pid)) unlinkMissingOkay(record.control_endpoint);
    const endpointClean = typeof record.control_endpoint !== "string" || !existsSync(record.control_endpoint);
    if (agentClean && workerClean && endpointClean) {
      this.closeRecord(record, "forced_stop");
      return { stopped: true, forced: true };
    }
    throw new BackendUnknownError("ACP stop could not prove Worker/Agent cleanup");
  }

  private closeRecord(record: RuntimeRecord, reason: string): void {
    stateStore.transaction((connection) => {
      const timestamp = stateStore.now();
      connection.execute(
        `UPDATE launches SET status='closed', prompt_state='cancelled',
          exit_reason=COALESCE(exit_reason, ?), closed_at=COALESCE(closed_at, ?), last_event_at=?
         WHERE launch_id=?`,
        [reason, timestamp, timestamp, record.launch_id],
      );
      connection.execute(
        "UPDATE acp_sessions SET status='closed', closed_at=COALESCE(closed_at, ?) WHERE launch_id=? AND status='active'",
        [timestamp, record.launch_id],
      );
    });
  }

  observe(options: { jobId?: string | null; sessionName?: string | null; cwd?: string | null } = {}): ObserveResult {
    const record = this.record();
    if (options.jobId && !new Set([record.backend_ref, AcpBackend.jobId(record)]).has(options.jobId)) {
      return new ObserveResult("unknown", null, "job id does not match Launch");
    }
    const workerAlive = pidAlive(record.worker_pid);
    const agentAlive = pidAlive(record.agent_pid);
    const workerGroupAlive = processGroupAlive(record.worker_pid);
    const agentGroupAlive = processGroupAlive(record.agent_pid);
    const endpointExists = typeof record.control_endpoint === "string" && existsSync(record.control_endpoint);
    const ping = endpointExists ? this.ping(record) : null;
    if (ping && workerAlive) return new ObserveResult("present", ping);
    if (!workerAlive && (agentAlive || agentGroupAlive)) {
      return new ObserveResult("unknown", { agent_pid: record.agent_pid }, "orphan Agent Process is alive without its ACP Worker");
    }
    if (!workerGroupAlive && !agentGroupAlive && !endpointExists) return new ObserveResult("absent");
    return new ObserveResult("unknown", null, "ACP Launch facts are contradictory");
  }

  listSessions(): RuntimeRecord[] {
    if (!this.executionRecord.root_id) return [];
    return stateStore.listLaunches(String(this.executionRecord.root_id))
      .filter((row) => row.backend_id === "acp" && !new Set(["closed", "error", "turn_ended"]).has(String(row.status)))
      .map((row) => ({
        id: AcpBackend.jobId(row), job_id: AcpBackend.jobId(row), launch_id: row.launch_id,
        name: row.session_name, session_name: row.session_name, state: row.status,
        worker_pid: row.worker_pid, agent_pid: row.agent_pid,
      }));
  }

  supportsHooks(): boolean { return false; }
}
